### PR TITLE
Chapter2

### PR DESCRIPTION
**Pull request by jamesthesnake**
### Solutions for Chapter 2
cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -37,7 +37,7 @@ want some feedback on your solutions.
 
 Now, if you are ready, bring it on!
 -}
-
+Jamesthesnake work
```
Comment (vrom911) [inline src/Chapter2.hs:40]: In order this to compile you could put this under the comment section 🙂

File: src/Chapter2.hs [hunk 2]
```diff
@@ -348,8 +350,10 @@ Implement a function that returns only the first half of a given list.
 >>> firstHalf "bca"
 "b"
 -}
--- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+firstHalf :: [a] -> [a]
+firstHalf l =
+    let halfLen = div (length l) 2
+    in take halfLen l
```
Comment (vrom911) [inline src/Chapter2.hs:356]: Great usage of `let-in` construction! Makes code easily readable 💯

File: src/Chapter2.hs [hunk 3]
```diff
@@ -620,7 +627,10 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven (x : _ : xs) = x : takeEven xs
```
Comment (vrom911) [inline src/Chapter2.hs:633]: Great pattern-matching skills 👏🏼

File: src/Chapter2.hs [hunk 4]
```diff
@@ -727,7 +737,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = concat (map (\x -> replicate x x) l)
```
Comment (vrom911) [inline src/Chapter2.hs:740]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 5]
```diff
@@ -740,7 +750,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Int -> [[Int]] -> [[Int]]
+contains x l = filter (elem x) l
```
Comment (vrom911) [inline src/Chapter2.hs:754]: Now, after you've mastered eta-reduction, you can apply such a technique to other functions as well  🙂 
But your implementation is already great 💯

File: src/Chapter2.hs [hunk 6]
```diff
@@ -857,10 +874,16 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind = go []
+  where
+    go :: [a] -> [a] -> [a]
+    go res [] = res
+    go res (x:xs) = go (x : res) xs
```
Comment (vrom911) [inline src/Chapter2.hs:882]: Amazing and clean work 🔥

**Review by vrom911 — APPROVED** [review]
Chapter Two is done ✔️ Congrats!

Very well done, and the advanced task is 💯

**Comment by jamesthesnake** [comment]
@chshersh  just need your approval

**Comment by jamesthesnake** [comment]
and hacktoberfest label

**Comment by vrom911** [comment]
@jamesthesnake one approval is enough in our projects, so you are fine. Also, we added the label as well. This PR can be closed 🙂

**Pull request closed without merging**